### PR TITLE
Multiple fixes

### PR DESCRIPTION
**Pull request by omniomi**
**General changes:**

- Double quoted strings are now a repository item (required for other fixes.)

**Scope changes:**

- Variable tokens (`$`,`@`) now scoped `keyword.other.variable.definition.powershell`. While they should be `punctuation.definition.variable` not many themes cover `punctuation.definition.*` due to there not being as many token based languages.

- Variables are now scoped `variable.other.readwrite.powershell`.

- Parentheses, brackets, and braces (`(`,`)`,`[`,`]`,`{`,`}`) moved to `Punctuation.section.*` scope.

- Types now scoped `storage.type.powershell`.

**Fixes:** 

- Attributes (like `[cmdletbinding()]` and `[validatepattern()]`)
  - Fixed highlighting issues where certain characters would break document highlighting.
  - Fixed parameter highlighting where certain spacing would break line highlighting.
  - Numerous scoping changes.

- Type declarations containing digits (like `[int32]`) no longer break highlighting.

attributes

### CODE REVIEW

**Comment by omniomi** [comment]
Fixes #75
Fixes #74
Fixes #73
Fixes #71
Fixes #65
Fixes #64
Fixes #59 - Theme -> Scope dependent.
Fixes #56
Fixes #51
Fixes #37
Fixes #38
Fixes #28
Fixes #27
Fixes #26
Fixes #17
Fixes #15
Fixes #14

**Comment by omniomi** [comment]
Just noticed, there's one edge case that breaks highlighting that I still need to figure out...

```PowerShell
[attribute(blah blah blah)
]
```

breaks highlighting even with a `\s*` between the `\)` and `\]` in the pattern for the end captures. 

Still working on...

**Comment by kborowinski** [comment]
This's really nice job @omniomi

**Comment by kborowinski** [comment]
@tylerl0706 I have noticed that *PowerShellSyntax.tmLanguage* didn't change in couple of latest *VSCode Insiders* builds. Should we ping someone at *VSCode* to updated it?

**Comment by omniomi** [comment]
@kborowinski can you post that snippet as plain text so I can copy/paste for testing please.

**Comment by kborowinski** [comment]
@omniomi  Yep, there you go:
```powershell
function Test-Highlighting {
    param(
        [ValidateScript({
            if (($_ -gt ($upperLimit = ((Get-Date) - (Get-Date '1601-01-01')).Days)) -or ($_ -lt 0)) {
                throw ("The specified value '{0}' is outside allowed range (0 to {1})" -f $_, $upperLimit)
            } else {
                $true
            }
        })]
        [Int]$InactivityDays = 0
    )
    $InactivityDays
}

```

**Comment by omniomi** [comment]
@kborowinski it's treating the closing brace of the if statement as the close of the scriptblock which puts "else" into meta.attribute.powershell which gives it no definition. 

The correct solution would be to properly define if/else/elseif statements not just their keywords... the quick and dirty solution is to include $self inside the attribute definition... I'm inclined to do it properly so it will take a bit.

ifelse-1

ifelse-2

**Comment by kborowinski** [comment]
@omniomi So enclosing the whole script in brackets should theoretically force parser to treat it as one block? Let me do quick check.

BTW, sorry for naive question, but how you get these nice popups with token details?

**Comment by tylerl0706** [comment]
@kborowinski I'll check to see what their process is exactly.

**Comment by tylerl0706** [comment]
Btw, just let me know when you're ready for me to review this :)

**Comment by omniomi** [comment]
@kborowinski, I addressed the if/else issue within the validate attribute. Should be good in the latest build: https://ci.appveyor.com/project/PowerShell/editorsyntax/build/1.0.42-yvpenuyl

Still needs to be fixed properly but it highlights correctly for now.

**Comment by kborowinski** [comment]
Nicely done!

image

**Comment by tylerl0706** [comment]
FYI https://github.com/Microsoft/vscode/issues/49789

**Comment by tylerl0706** [comment]
FYI VSCode updated to the latest master, we should see some changes in the next day or so :)

**Comment by omniomi** [comment]
@tylerl0706 this should be ready by tonight. The dependency list won't change:

- **Node.js** (already dealt with)
- **fast-plist** (already dealt with, made by Microsoft.)
- **Atom.io** (for tests)
- **atom-grammar-test** (for tests) https://github.com/kevinastone/atom-grammar-test

Can you do whatever  needs to be done internally for the Atom and atom-grammar-test requirements?

**Comment by tylerl0706** [comment]
Yeah I'll take care of that

File: build.ps1 [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        Finished in 0.281 seconds
+        2 tests, 5 assertions, 0 failures, 0 skipped
+#>
+[cmdletbinding()]
+param(
+    [parameter()]
+    [ValidateSet('Build','Test')]
+    [string]
+    $Task = 'Build'
```
Comment (tylerl0706) [inline build.ps1]: Do you think you could modify this build script to be more consistent to the other repo's build.ps1?

Take a look at https://github.com/PowerShell/Polaris/blob/master/build.ps1

```powershell
param(
    [Parameter()]
    [switch]
    $Bootstrap,  # prints out what needs to be installed

    [Parameter()]
    [switch]
    $Test,

    [Parameter()]
    [switch]
    $Package 
)
```

and then running it by itself will build.

**Comment by omniomi** [comment]
@kborowinski I had a look at enum coverage which working through the specs. It's surprisingly difficult (without running into conflicts with other scopes.) I won't be able to include it in this batch of fixes but we will get it done.

**Comment by omniomi** [comment]
@tylerl0706, @kborowinski, I consider this guy ready to go.

It doesn't fix _everything_ but as a unit it's complete and introduces no regression.

**Comment by omniomi** [comment]
@tylerl0706 

**Screenshots**

screenshot1

- `--%` now treated as a singular keyword not as `--` (subtract) and `%` (foreach)
- Variable tokens now scoped correctly.

screenshot2

- Digits in types don't break highlighting.
- Types scoped correctly.

screenshot3

- 'default' keyword highlights correctly.
- Braces, brackets, and parens all now correctly scoped as punctuation and using 'begin' and 'end'. 

screenshot4

- Comments between function name and K&R style brackets highlight correctly.
- More examples of consistent section open/close punctuation.

screenshot5
screenshot6

- Strings no longer broken by characters in regex.

screenshot7

- cmdlets in attributes no longer break highlighting on dash.

**Other changes** are either about consistency and are difficult to capture in screenshots, are minor rescoping changes, or are related to the development of the CI tests / scripts.

File: tools/build-helpers.ps1 [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    exit
+}
+
+function RunSpecs {
+    Write-Host "Running specs..."
+
+    $specpath = Join-Path . '\spec'
+
+    if (Test-Path $specpath) {
+      & "$script:ATOM_EXE_PATH" --test spec *>&1 | ForEach-Object { ParseJasmine $_ }
```
[the remark this answers is not in the document]
Comment (omniomi) [reply tools/build-helpers.ps1]: @vors That stuff is all ripped from https://raw.githubusercontent.com/atom/ci/master/build-package.ps1 I am not married to it. I'd rather redesign the CI in a new PR once this is merged than bloat this one further though.

File: .gitignore [hunk 3]
```diff
@@ -1,2 +1,5 @@
 syntaxes/
-tools/node_modules/
\ No newline at end of file
+grammars/
+node_modules/
+atom/
+atom.zip
```
[the remark this answers is not in the document]
Comment (omniomi) [reply .gitignore]: Nevermind, Google told me. Will do.

**Comment by omniomi** [comment]
Snuck in one more little fix, email addresses within "strings" were highlighting like a splatted variable.

@tylerl0706 she's good to go.

File: appveyor.yml [hunk 4]
```diff
@@ -20,13 +20,24 @@ skip_commits:
...
+  - ps: SyntaxBanner
   - ps: Install-Product node '8.9.1'
-  - ps: Set-Location (Join-Path $env:APPVEYOR_BUILD_FOLDER '\tools\')
+  - ps: Set-Location $env:APPVEYOR_BUILD_FOLDER
   - ps: npm install
+  - ps: DownloadAtom
 
 build_script:
-  - ps: Set-Location (Join-Path $env:APPVEYOR_BUILD_FOLDER '\tools\')
+  - ps: BuildBanner
```
Comment (tylerl0706) [inline appveyor.yml]: You don't have to address this now but I want to mention that I'm hoping to try and make the build.ps1 generic enough that we can use just build.ps1 commands in the appveyor.yml.

Then we can easily drop in an appveyor.yml into any repo that follows the build.ps1 concept and we'll have CI :)

**Comment by vors** [comment]
🎉

**Pull request merged**